### PR TITLE
[SE-0274] Update concise #file implementation in response to first round of review

### PR DESCRIPTION
**Pull request by beccadax**
This PR addresses several issues that came up in the first review of SE-0274:

* The format of the `#file` string is now `<module>/<filename>`. (Technically, the proposal will also allow it to be `<module>/<other-stuff>/<filename>` instead, but that's for future expansion.)

* If a `#sourceLocation(file:line:)` directive maps to the same `#file` string as a real file or another `#sourceLocation` directive, but the paths don't match exactly, the compiler now emits a warning.

* `-emit-sil` and related textual SIL outputs now include a comment with a table mapping `#file` strings to `#filePath` strings. This is a proof-of-concept demonstrating that we can expose this information in e.g. SourceKit later on.

The proposal will be updated to include these changes shortly.

### CODE REVIEW

**Comment by brentdax** [comment]
@CodaFi This set of changes works, but I'm not very happy with where this code lives in the compiler or how it ends up being called. Do you have any suggestions to improve it?

File: include/swift/AST/SourceFile.h [hunk 1]
```diff
@@ -230,6 +230,12 @@ class SourceFile final : public FileUnit {
   /// forwarded on to IRGen.
   ASTStage_t ASTStage = Parsing;
 
+  /// Virtual filenamess declared by #sourceLocation(file:) directives in this
```
Comment (brentdax) [inline include/swift/AST/SourceFile.h]: Typo; I should fix this.

**Comment by brentdax** [comment]
@swift-ci test

**Comment by brentdax** [comment]
@swift-ci test Windows platform

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test Windows platform

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please test windows platform

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please test windows platform

**Comment by brentdax** [comment]
@swift-ci please test windows platform

**Review by hamishknight — COMMENTED** [review]
Looks great! Just a couple of nits

**Pull request merged**